app/services/katello/ui_notifications/hosts/lifecycle_expire_soon.rb

```diff
@@ -0,0 +1,58 @@
+module Katello
+  module UINotifications
+    module Hosts
+      class LifecycleExpireSoon
+        def self.deliver!
+          ::Katello::RhelLifecycleStatus.lifecycles_expire_soon.each do |release, schedule|
+            schedule.each do |lifecycle, end_date|
+              count = hosts_with_index(release).count
+              next if count == 0
+
+              message = message(count: count, release: release, lifecycle: lifecycle, end_date: end_date)
+              if (notification = existing_notification(release))
+                /[^:]+: (?<number_of_hosts>\d+) hosts/ =~ notification.message
+                next if number_of_hosts == count.to_s
+                notification.update(message: message)
+              else
+                ::Notification.create!(
+                  :initiator => User.anonymous_admin,
+                  :audience => Notification::AUDIENCE_GLOBAL,
+                  :message => message,
+                  :expired_at => end_date.strftime('%Y-%m-%d'),
+                  :notification_blueprint => blueprint
+                )
+              end
+            end
+          end
+        end
+
+        def self.existing_notification(release)
+          blueprint.notifications.where("message like ?", "#{release}%").first
+        end
+
+        def self.message(options)
+          ::UINotifications::StringParser.new(
+            blueprint.message,
+            :number_of_hosts => options[:count],
+            :release => options[:release],
+            :lifecycle => options[:lifecycle].gsub(/_/, " "),
+            :end_date => options[:end_date].strftime('%Y-%m-%d'),
+            :audience => Notification::AUDIENCE_GLOBAL
+          )
+        end
+
+        def self.hosts_with_index(release)
+          /RHEL(?<major>\d+)/ =~ release
```

Comment:
Is this what assigns a value to `major`?

Reply:
That is right. [This post](https://medium.com/rubycademy/3-ways-to-use-regexp-capture-groups-with-back-references-in-ruby-b4969cc9b3ec) explaines it well.